src/main/java/de/tum/in/www1/artemis/service/hestia/TestwiseCoverageService.java

```diff
@@ -0,0 +1,233 @@
+
+    private final CoverageFileReportRepository coverageFileReportRepository;
+
+    private final TestwiseCoverageReportEntryRepository testwiseCoverageReportEntryRepository;
+
+    private final ProgrammingExerciseTestCaseRepository programmingExerciseTestCaseRepository;
+
+    private final ProgrammingSubmissionRepository programmingSubmissionRepository;
+
+    private final GitService gitService;
+
+    private final RepositoryService repositoryService;
+
+    private final SolutionProgrammingExerciseParticipationRepository solutionProgrammingExerciseParticipationRepository;
+
+    public TestwiseCoverageService(CoverageReportRepository coverageReportRepository, CoverageFileReportRepository coverageFileReportRepository,
+            TestwiseCoverageReportEntryRepository testwiseCoverageReportEntryRepository, ProgrammingExerciseTestCaseRepository programmingExerciseTestCaseRepository,
+            ProgrammingSubmissionRepository programmingSubmissionRepository, RepositoryService repositoryService,
+            SolutionProgrammingExerciseParticipationRepository solutionProgrammingExerciseParticipationRepository, GitService gitService) {
+        this.coverageReportRepository = coverageReportRepository;
+        this.coverageFileReportRepository = coverageFileReportRepository;
+        this.testwiseCoverageReportEntryRepository = testwiseCoverageReportEntryRepository;
+        this.programmingExerciseTestCaseRepository = programmingExerciseTestCaseRepository;
+        this.programmingSubmissionRepository = programmingSubmissionRepository;
+        this.gitService = gitService;
+        this.repositoryService = repositoryService;
+        this.solutionProgrammingExerciseParticipationRepository = solutionProgrammingExerciseParticipationRepository;
+    }
+
+    /**
+     * Creates a coverage report from a testwise coverage report.
+     * Test case names are resolved to a test case of the given programming exercise, adds this reference to the given
+     * entries and saves the entries with the test case reference to the database.
+     * In case, no test case could be found for the given name, the report for this test case will not be saved
+     * @param fileReportByTestCaseName a map containing the test case name as a key and the file coverage reports without the
+     * reference to a test case as a value
+     * @param submission the solution programming submission for which the report is updated
+     * @param exercise the exercise for which the report should be updated
+     */
+    public void createTestwiseCoverageReport(Map<String, Set<CoverageFileReport>> fileReportByTestCaseName, ProgrammingExercise exercise, ProgrammingSubmission submission) {
+        // If the report already exists, do not create a new report. This is the case if the build plan will be re-run
+        boolean reportAlreadyExists = coverageReportRepository.findCoverageReportBySubmissionId(submission.getId()).isPresent();
```

Comment:
If you really want to check if the report exists already I would suggest one of two things:
Either create a specific query that does not fetch the entire Submission with all parent objects (this would be unnecessarily computationally expensive here) or ensure that the submission that is passed into this method already has the coverage report fetched.

Reply:
I will go with the first option. The submission will never have the report fetched because it is to be created inside for this method for the given submission.